### PR TITLE
[WIP] Two factor authentication

### PR DESCRIPTION
**Pull request by jbrooksuk**
This is a biggy for #326 and is a massive WIP.

- [x] Add package.
- [x] Setup Google 2FA Secret on setup.
- [x] Show QR code in user panel.
- [x] Enable/disable 2FA in user panel.
- [x] If user logs in and has 2FA enabled, show code form (currently working on).
- [x] Language file updates.

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Phew. Two factor authentication is implemented. Can a couple of you try it out please? I've been able to log in normally without two-factor turned on and if I turn it on, I can only login by supplying a token.

@cachethq/owners FYI.

**Comment by joecohens** [comment]
@jbrooksuk Testing now ...

**Comment by joecohens** [comment]
@jbrooksuk It's working nice work! But I think the flow is not the best because you can get locked out easily.

I'll remove the enable two factor auth from the setup page and leave it only on the profile. If we leave the button we have to redirect after login to setup MFA code directly. I think some people wont know what to do after checking the box. 

Also when you go to the teams page and you see yourself we can setup a button like edit your profile or something to remember people how to.

What do you think?

File: app/views/dashboard/user/index.blade.php [hunk 1]
```diff
@@ -42,6 +42,27 @@
...
                         </div>
+                        <hr />
+                        <div class="form-group">
+                            <label class="checkbox-inline">
+                                <input type="hidden" name="google2fa" value="0" />
+                                <input type='checkbox' name="google2fa" value="1" {{ !empty(Auth::user()->google_2fa_secret) ? "checked" : "" }} />
+                                {{ trans('forms.setup.enable_google2fa') }}
+                            </label>
+                        </div>
+                        @if(Auth::user()->google_2fa_secret)
```
Comment (joecohens) [inline app/views/dashboard/user/index.blade.php]: Can we make this validation on the user like:

Model\User
````
function hasEnabled2FA()
{
    return (bool) (trim($this->google_2fa_secret) !== '');
}
````

Comment (jbrooksuk) [inline app/views/dashboard/user/index.blade.php]: Good idea.

**Comment by jbrooksuk** [comment]
@joecohens good point. If you enable it, but then don't get the QR code you can indeed lock yourself out.

Can you push those suggestions? I'm not able to right now.

**Comment by joecohens** [comment]
@jbrooksuk Sure!

**Comment by jbrooksuk** [comment]
Done. Time to rebase and then merge!

**Comment by jbrooksuk** [comment]
Also we need to move the code for the QR code into the controller. It shouldn't be in the view.

File: app/config/app.php [hunk 2]
```diff
@@ -172,17 +173,17 @@
...
-        'Form'     => 'Illuminate\Support\Facades\Form',
-        'Input'    => 'Illuminate\Support\Facades\Input',
-        'Redirect' => 'Illuminate\Support\Facades\Redirect',
-        'Request'  => 'Illuminate\Support\Facades\Request',
-        'Response' => 'Illuminate\Support\Facades\Response',
-        'Route'    => 'Illuminate\Support\Facades\Route',
-        'Session'  => 'Illuminate\Support\Facades\Session',
-        'Setting'  => 'CachetHQ\Cachet\Models\Setting',
-        'Str'      => 'Illuminate\Support\Str',
+        'App'       => 'Illuminate\Support\Facades\App',
```
Comment (GrahamCampbell) [inline app/config/app.php]: this indentation should not have changed.

Comment (jbrooksuk) [inline app/config/app.php]: Whoops. I did that when I originally added the new library.

File: app/views/auth/two-factor-auth.blade.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
+@extends('layout.clean')
+
+@section('content')
+    <div class="login">
+        <div class="col-xs-12 col-xs-offset-0 col-sm-6 col-sm-offset-3 col-md-4 col-md-offset-4 text-center">
+            <div class="welcome-logo">
+                <img class="logo" height="50" src="{{ url('img/cachet-logo.svg') }}" alt="Cachet"/>
```
Comment (GrahamCampbell) [inline app/views/auth/two-factor-auth.blade.php]: you should use the "asset" helper for assets

File: app/views/dashboard/team/index.blade.php [hunk 4]
```diff
@@ -21,7 +21,7 @@
                 <div class="user-grid">
                     @foreach($teamMembers as $member)
                     <div class="user col-sm-3 col-xs-6">
-                        <a href="/dashboard/team/{{ $member->id }}">
+                        <a href="@if(Auth::user()->id == $member->id) {{ url('dashboard/user') }} @else /dashboard/team/{{ $member->id }}  @endif">
```
Comment (GrahamCampbell) [inline app/views/dashboard/team/index.blade.php]: too many spaces before the closing endif

File: app/views/auth/two-factor-auth.blade.php (hunk 3 above)
Comment (jbrooksuk) [inline app/views/auth/two-factor-auth.blade.php]: The `asset` url breaks things when you're on HTTPS.

**Comment by jbrooksuk** [comment]
Leaving for @joecohens to fix, I've gotta leave for a bit,

File: src/Http/Controllers/SetupController.php [hunk 5]
```diff
@@ -126,12 +126,11 @@ public function postStep2()
             // Pull the user details out.
             $userDetails = array_pull($postData, 'user');
 
-            // TODO: Do we want to just use Model::unguard() here?
             $user = User::create([
-                'username' => $userDetails['username'],
-                'email'    => $userDetails['email'],
-                'password' => $userDetails['password'],
-                'level'    => 1,
+                'username'          => $userDetails['username'],
```
Comment (GrahamCampbell) [inline src/Http/Controllers/SetupController.php]: incorrect indentation

File: src/Models/User.php [hunk 6]
```diff
@@ -133,4 +133,14 @@ public function getIsAdminAttribute()
...
     }
+
+    /**
+     * Returns if a user has enabled two factor authentication.
+     *
+     * @return bool
+     */
+    public function getHasEnabled2FAAttribute()
+    {
+        return (bool) (trim($this->google_2fa_secret) !== '');
```
Comment (GrahamCampbell) [inline src/Models/User.php]: no need to cast this to a bool

File: app/views/auth/two-factor-auth.blade.php (hunk 3 above)
Comment (GrahamCampbell) [inline app/views/auth/two-factor-auth.blade.php]: That's a bummer. :(

Comment (GrahamCampbell) [inline app/views/auth/two-factor-auth.blade.php]: It's also odd. That helper function works for me on styleci and on bootstrap cms, using L4.2 and L5.0.

Comment (joecohens) [inline app/views/auth/two-factor-auth.blade.php]: Yeah I saw you where using the `assets` on StyleCI.

**Comment by joecohens** [comment]
This is done :) the only thing missing is the suggestion of @GrahamCampbell to change te `getHasEnabled2FAAttribute` method name.

**Comment by jbrooksuk** [comment]
Does this stop the user password from being changed too?

**Comment by joecohens** [comment]
@jbrooksuk Now it does :)

**Comment by jbrooksuk** [comment]
@joecohens what's left?

**Comment by jbrooksuk** [comment]
So we should be able to login with oAuth, disable/enable it and it will no longer lose the password?

**Comment by joecohens** [comment]
@jbrooksuk Yes!

**Pull request merged**